constraint-solver/src/solver/base.rs

```diff
@@ -219,18 +219,57 @@ where
     /// If there is exactly one assignment that does not lead to a contradiction,
     /// apply it. This might be expensive.
     fn exhaustive_search(&mut self) -> Result<bool, Error> {
-        let assignments = exhaustive_search::get_unique_assignments(
-            self.constraint_system.system(),
-            &*self,
-            &self.bus_interaction_handler,
-        )?;
+        log::debug!("Starting exhaustive search...");
+        let mut variable_sets =
+            exhaustive_search::get_brute_force_candidates(self.constraint_system.system(), &*self)
+                .collect_vec();
+        // Start with small sets to make larger ones redundant after some assignments.
+        variable_sets.sort_by_key(|set| set.len());
+
+        log::debug!(
+            "Found {} sets of variables with few possible assignments. Checking each set...",
+            variable_sets.len()
+        );
 
         let mut progress = false;
-        for (variable, value) in &assignments {
-            progress |=
-                self.apply_range_constraint_update(variable, RangeConstraint::from_value(*value));
-        }
 
+        let mut error_sets = BTreeSet::new();
+
+        for mut variable_set in variable_sets {
+            variable_set.retain(|v| {
+                self.range_constraints
+                    .get(v)
+                    .try_to_single_value()
+                    .is_none()
+            });
+            if error_sets.contains(&variable_set) {
```

Comment:
I assume this is a performance optimization? So what happens is:
1. A smaller set is processed, leads to assignments
2. Another smaller set does not lead to assignments, gets added to the error set.
3. A larger set becomes smaller by the filter step above, and is equal to set (2). Then we skip it, because we already processed it?

Reply:
Yes exactly that is what I had in mind. I don't know if there are the exact steps, but in sha256 the same variable set is tried multiple times without success.